src/waitress/channel.py

```diff
@@ -194,7 +226,7 @@ def received(self, data):
             data = data[n:]
 
         if requests:
-            self.requests = requests
+            self.requests.extend(requests)
```

Comment:
This becomes very unfair, if someone were to keep pipelining requests, the task thread that is servicing the requests will get stuck inside it's while self.requests loop with the task lock held... meaning that someone pipelining one request after the other could stop any other requests from moving forward because they would never get scheduled.

Reply:
If I understand the code correctly, the task lock is defined for each channel and each channel is only responsible for one socket, so different connections should not influence each other. If one client keeps pipelining requests, `len(self.requests)` will become larger than `max_requests_per_channel` and no more `read` is performed until enough requests have been answered.

Is this any different from the behavior before if a client kept writing more and more requests into the same connection?

Reply:
On the line right after this you re-add the task to the queue of items to be processed. This means that if thread 1 is already processing all requests (and thus has the task lock), another thread may pick up that it needs to work on this channel because it was added to the queue and it will wait until it can get the task lock.

For each request that is added it will block. Each time a request is popped off, dropping below the maximum another request may be read.

Reply:
Each time you call `self.server.add_task(self)` it adds the current channel to the list of items to be processed.

However only a single thread can hold the task lock. If we keep adding to `self.requests` that single thread that is processing the current request while we read this one will continue to process requests.

In the mean time each time we added ourselves to the backlog on line 216 we could potentially get scheduled and block waiting for us to get the task lock.

You'd need to make sure you only schedule a task if its not already running. You'd want to do a non-blocking acquire, and only if you acquired the lock should you add yourself to the task queue... although I foresee that if multiple reads are allowed but the threads are really busy you'd still potentially schedule the same channel multiple times, which could still lead to the same deadlock.

Reply:
Instead we'd probably want to rewrite waitress so that if there are multiple requests, the channel reschedules itself for execution at the end of the line, rather than trying to process all of the requests in the queue. This way pipelined requests are scheduled more fairly.

All in all that is a much larger change, since requests have to be processed in the order received because the client expects responses in the order it sent them.

Reply:
Right, thanks, now I understand. Only one thread will ever execute for a given channel due to the `task_lock`, but it might happen that one channel blocks all threads that `waitress` has available for tasks to be finished, while only one of them does any actual work.

So yes, it seems that `service` should only `pop` one request to work on it and, just before finishing, call `add_task` again if there are more requests. I will try to think about how to accomplish this without race conditions.

Reply:
I believe this has been resolved now, as the while loop as been removed and `channel.service()` only handles one request at a time, ensuring to invoke `server.add_task(self)` if there are more requests to execute keeping things fair.